### ISSUE TITLE
Add a Pfsense package

### ISSUE THREAD

**Issue by mrash** [mrash/fwknop#130]
Add support for Pfense firewalls with a dedicated package. FreeBSD is already supported, so this should be relatively easy.

**Comment by JeffMelton** [comment]
Are you still considering this issue? I'm not able to support with a PR, but I'm a pretty good guinea pig if you need a tester.

**Comment by mrash** [comment]
Yes, will definitely do this, probably for the 2.6.8 release I think.

**Comment by RusteeGit** [comment]
Yes!! Such a good idea! Everyone should be using fwknop and this would make it so usable!

**Comment by Joeviocoe** [comment]
I am looking forward to this package for pfsense (2.2.5+).
Would I need to compile on a freebsd vm to make the package first?  Or will fwknop 2.6.8 be ready for pkg?

Let me know if you need any help testing.  Thanks.

**Comment by Joeviocoe** [comment]
Did 2.6.8 get support for pfsense?

**Comment by mrash** [comment]
Not yet, haven't had time to work on this yet. Seems like this should be a priority though.

**Comment by cwross** [comment]
Yes! This would be so cool! Thank you for your efforts!

**Comment by Joeviocoe** [comment]
Just checking in if this has become a priority :)

**Comment by Joeviocoe** [comment]
I am sure life gets in the way.  So not trying to nag.
But is there any ETA on fwknop-2.6.10 and its milestone features such as pfsense support?
Thanks.

**Comment by stilez** [comment]
Two quick notes:

1) pfSense uses pf on FreeBSD. The docs say that fwknop supports pf on OpenBSD or ipfw on FreeBSD. So it might be that work is needed to make it compatible with pf on FreeBSD (if it isn't already) or the docs need updating (if it is). I've added FreeBSD pf compatibility as a distinct issue (#269)

2) If fwknop is (or becomes) compatible with pf on FreeBSD, I'm a keen pfSense developer and if not too busy at the time, I'd be happy to see if I can help out on making this available. Contact (mildly scrambled): stilezy-att-gm-ai_ldott_(com).

**Comment by Joeviocoe** [comment]
I grew a bit impatient, please forgive me.

I don't have the lower level coding skill do help with this particular project, but I really really wanted to protect my Internet facing service with something.

I do know bash, so here's what I whipped up over a few days:

Android (not rooted), runs Tasker automation, but ultimately is just calling a shell in one action.  Also, any authentication app such as Authy or Google Authenticator.  Tasker clears clipboard, opens authenticator app, and waits for user to copy OTP token (6 digits changing every 30 seconds).  When copied to clipboard, Tasker sends shell command "One ping only".
Ping command is padded with unique static string, 2 control bytes, and the OTP token.

Pfsense runs a bash script.  Running tcpdump on a loop, with the -c argument ending the command after X icmp packets received, then waiting 2X seconds to restart (prevents brute force).  Filters on tcpdump look for that unique static string inside the ICMP echo payload, then passes the payload data to awk which extracts the OTP token and the control bytes and passes it to a 2nd bash script.  This next bash script does the HMAC magic using openssl and other binaries already included in pfsense.  Using the shared secret that is also in the client's authenticator app, it confirms the same OTP for that time interval of 30 seconds.  The only thing I needed to install on pfsense, is a base32 module for perl which isn't preinstalled (MIME::base32).  Once OTP token is confirmed, the script will pass the 2 control bytes to a 3rd bash script that simply uses the control bytes to execute whatever command is predetermined.  One byte is On or Off, the other byte is the index of a service or firewall rule.

This **Ping Knock** method is not as good as what fwknop is going to bring to Pfsense... mostly because it is limited to unencrypted payloads and the small key that TOTP tokens use for a secret.  But this is very good for LOL (Living off the land) as it uses the encryption and binaries we already have.  

It is a single, small packet, most IDS won't alert on a ping, and resists brute force.  It mitigates replay attacks, well the attacker has 30 seconds.  
I haven't implemented passing the ICMP source IP address to execute a firewall rule which includes src addr... but that is easy since it's already part of the ICMP packet.  Of course that opens it up to MITM manipulation (trusting the header).  But for just starting a service, rather than setting up user authorization with src IP... it works great for now.

Please continue your good work on the pfsense implementation of fwknop.  I look forward to it.  Also, I hope the fwknop2 android client will have Tasker integration so I can send the SPA in the background based on whatever rule I choose.

Thank you.

**Comment by securenotebook** [comment]
Hi Joe, any chance you have a more detailed write-up on this? or made any other progress?

**Comment by JamesTheAwesomeDude** [comment]
Has sense already handled this upstream?

https://github.com/pfsense/FreeBSD-ports/tree/master/security/fwknop
